engine/resources/hetzner.go

```diff
@@ -0,0 +1,710 @@
+	obj.rescueEnable = (obj.ServerRescueMode != ServerRescueDisabled)
+
+	return nil
+}
+
+// Close runs some cleanup code for this resource.
+// TODO: clear the root password info if rescue mode was enabled?
+func (obj *HetznerRes) Close() error {
+	obj.client = nil
+	obj.server = nil
+	return nil
+}
+
+// Watch is not implemented for this resource, since the Hetzner API does not
+// provide any event streams. Instead, always use polling.
+// Note that Validate sets an explicit minimum on the polling interval.
+func (obj *HetznerRes) Watch() error {
+	return fmt.Errorf("invalid Watch call: HetznerRes requires polling")
+}
+
+// CheckApply checks the resource state and applies the resource if the bool
+// input is true. It returns error info and if the state check passed or not.
+func (obj *HetznerRes) CheckApply(apply bool) (bool, error) {
+	var checkOk = true
+	ctx := context.TODO()
+	// Request up-to-date server info from the API.
+	if err := obj.getServerUpdate(ctx); err != nil {
+		return false, errwrap.Wrapf(err, "getServerUpdate failed")
+	}
+	// Check if the server instance is in the correct state.
+	if c, err := obj.checkApplyServerState(ctx, apply); err != nil {
+		return false, errwrap.Wrapf(err, "")
+	} else if !c {
+		checkOk = false
+	}
+	// Skip further steps if not in the intended state.
+	if !apply && !checkOk {
+		return false, nil
+	}
+	// Skip further steps if the server is not running.
+	if obj.State != StateRunning {
+		return checkOk, nil
+	}
+	// Changes in cpu, location and/or image require a server rebuild.
+	if c, err := obj.checkApplyServerRebuild(ctx, apply); err != nil {
+		return false, errwrap.Wrapf(err, "checkApplyServerRebuild failed")
+	} else if !c {
+		checkOk = false
+	}
+	// Changes in rescue mode can be made while the server is running.
+	if c, err := obj.checkApplyRescueMode(ctx, apply); err != nil {
+		return false, errwrap.Wrapf(err, "checkApplyRescueMode failed")
+	} else if !c {
+		checkOk = false
+	}
+	// TODO: checkapply SSH keys
+	return checkOk, nil
+}
+
+// checkApplyServerState powers the server on/off, creates a new instance, or
+// deletes the existing one as needed, to ensure the correct state.
+func (obj *HetznerRes) checkApplyServerState(ctx context.Context, apply bool) (bool, error) {
+	if obj.init.Debug {
+		obj.init.Logf("checkApplyServerState(apply: %t)", apply)
+	}
+	var checkOk = true
+	// Make sure the server exists as intended before further checks.
+	// TODO: add create permission
+	if obj.server == nil {
+		if obj.State == StateAbsent {
+			return true, nil
+		}
+		if !apply {
+			return false, nil
+		}
+		checkOk = false
+		obj.init.Logf("requesting server creation")
+		if err := obj.createServer(ctx); err != nil {
+			return false, errwrap.Wrapf(err, "createServer failed")
+		}
+	}
+	// Make sure the server is in a steady state before further changes.
+	// TODO: test if hetzner api accepts queries while in transient state
+	if err := obj.waitForSteadyState(ctx); err != nil {
```

Comment:
how long can this block for? is it guaranteed to eventually unblock? do we have a constant for a maximum wait time? or a param?

Reply:
Agreed, this needs something to exit if it keeps running for too long. It doesn't have a limit yet, but a maximum wait time sounds reasonable. I will add a param with a default value of 10 times the wait interval, that seems reasonable. Let me know if you have counterproposals.

Reply:
I think you probably just want to have a const at the very top with some super long, but not overly long time... 10 min perhaps?? Basically something that should never happen unless hetzner is broken... This way, we'll eventually make progress even if the mgmt user waits 10 min, they'll at least get an error... Blocking forever is bad. Make sense?

Reply:
The default is set to five minutes now, should be alright.